Implemetation of ability crosshairs

Badulundabad opened the pull request:
Added new ability effect types (TargetedStatic, Summoning);
Added SummoningScript class;
Some strings about entity frame were moved from Ability.ActivateAbility method to Ability.GetSpawnFrame method;
Ability.IsStaticAbility method was reworked to IsDynamicAbility for better using;
Removed AbilityScript.ShouldMove method, now Ability.IsDynamicAbility method is using instead;
Excess strings were removed from AbilityScript.UpdatePosition method;
Reworked AbilityManagerMissionLogic;
Changed key bindings, now it works so:
Q - enable/disable ability mode;
MouseScrollUp/MouseScrollDown - change ability;
LeftMouseButton - cast ability.

Badulundabad committed:
Added ability crosshairs

Badulundabad committed:
Adjusted ability crosshair logics

martkartasev commented on CSharpSourceCode/Abilities/Ability.cs:
```diff
@@ -93,50 +96,69 @@ public virtual void ActivateAbility(Agent casterAgent)
+                frame = UpdateFrameRotationForAI(casterAgent, frame);
+            //if (IsGroundAbility())
+            //    frame.origin.z = Mission.Current.Scene.GetGroundHeightAtPosition(frame.origin);
+            //frame = frame.Advance(Template.Offset);
```
There is a lot of commented code in the above block. Please clean it up.

martkartasev commented on CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:
```diff
@@ -1,31 +1,79 @@
+            }
+            else if (!isMainAgentChecked && Agent.Main != null)
+            {
+                TOWCommon.Say("YEP");
```
Should remove debugging code.

martkartasev commented on CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:
```diff
@@ -42,5 +90,48 @@ public override void OnAgentCreated(Agent agent)
+            offHand = Agent.Main.GetWieldedItemIndex(Agent.HandIndex.OffHand);
+            Agent.Main.TryToSheathWeaponInHand(Agent.HandIndex.MainHand, Agent.WeaponWieldActionType.WithAnimationUninterruptible);
+            sheathTimer.Start();
+            TOWCommon.Say("Ability mode is enabled");
```
Remove logging.

martkartasev commented on CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:
```diff
@@ -42,5 +90,48 @@ public override void OnAgentCreated(Agent agent)
+            ChangeKeyBindings();
+            Agent.Main.TryToWieldWeaponInSlot(mainHand, Agent.WeaponWieldActionType.WithAnimationUninterruptible, false);
+            Agent.Main.TryToWieldWeaponInSlot(offHand, Agent.WeaponWieldActionType.WithAnimationUninterruptible, false);
+            TOWCommon.Say("Ability mode is disabled");
```
Remove logging.

martkartasev commented on CSharpSourceCode/Abilities/Scripts/SummoningScript.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace TOW_Core.Abilities.Scripts
+{
+    public class SummoningScript : AbilityScript
```
Beginning of something? Remove for now?

hunharibo replied:
its how we did with the others too, even if empty. The factory needs it to assign behaviour, even if it does nothing compared to the basic AbilityScript. I guess we could just return AbilityScript in the factory, but i like it like this :)

martkartasev replied:
Fair enough. Its easier to reason about spell types in behaviors this way as well.

martkartasev reviewed (changes requested):
Review ability class.
AI Frame spawning wont work with crosshair.

I would suggest separate methods for AI frame and Player frame spawning.

Badulundabad committed:
Fixed GetSpawnFrame method

Badulundabad committed:
removed debug marker from DirectionalAOE

Badulundabad committed:
Fixed issue with invisible crosshair

Badulundabad committed:
Fixed issue with using abilities after death

hunharibo reviewed (approved):
Works for me. Thanks for taking care of the last bug. Approved.

Z3rca commented on CSharpSourceCode/Abilities/Ability.cs:
```diff
@@ -93,49 +96,85 @@ public virtual void ActivateAbility(Agent casterAgent)
+        private bool IsDynamicAbility()
         {
-            return Template.AbilityEffectType == AbilityEffectType.TargetedStaticAOE || Template.AbilityEffectType == AbilityEffectType.CenteredStaticAOE;
+            return Template.AbilityEffectType == AbilityEffectType.DynamicProjectile ||
```
i don't know I would prefer a case switch here. you can maybe sneak in here then an internal event we can register on

Badulundabad replied:
Yep, why not

Z3rca commented on CSharpSourceCode/Abilities/Ability.cs:
```diff
@@ -93,49 +96,85 @@ public virtual void ActivateAbility(Agent casterAgent)
+                    }
+                    else
+                    {
+                        frame = frame.Elevate(_template.Radius / 2);
```
maybe you should make a reference instead of a magic number.

Badulundabad replied:
hmmm, this code from ability implementation work, not mine, but i can change it, for example to add new field in AbilityTemplate

Z3rca commented on CSharpSourceCode/Abilities/Ability.cs:
```diff
@@ -93,49 +96,85 @@ public virtual void ActivateAbility(Agent casterAgent)
-            if (_template.AbilityEffectType == AbilityEffectType.CenteredStaticAOE)
-            {
-                frame = casterAgent.AgentVisuals.GetGlobalFrame();
+                if (_template.AbilityEffectType == AbilityEffectType.MovingProjectile || _template.AbilityEffectType == AbilityEffectType.DynamicProjectile)
```
a good case for a switch case in my opinion

Badulundabad replied:
yep

Z3rca commented on CSharpSourceCode/Abilities/Ability.cs:
```diff
@@ -185,48 +224,42 @@ private void AddLight(ref GameEntity entity)
 
         private void AddPhysics(ref GameEntity entity)
         {
-            if (!IsStaticAbility()) 
```
you do the same thing for both cases?

Badulundabad replied:
I think we should check if the ability is dynamic before run this method

Z3rca replied:
but you do the same thing irrespective what you choose... or did I read wrong

Badulundabad replied:
yes, it's the same thing, but i think it looks more logically

Z3rca commented on CSharpSourceCode/Abilities/AbilityFactory.cs:
```diff
@@ -37,35 +36,56 @@ public static void LoadTemplates()
+
+        private static AbilityCrosshair InitializeCrosshair(AbilityTemplate template, Agent caster)
+        {
+            AbilityCrosshair crosshair = null;
```
case switch maybe?

Badulundabad replied:
Yeah, Switch-Case commit awaits us)

Z3rca commented on CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:
```diff
@@ -42,5 +89,46 @@ public override void OnAgentCreated(Agent agent)
+        {
+            if (abilityComponent.IsAbilityModeOn)
+            {
+                keyContext.GetGameKey(11).KeyboardKey.ChangeKey(InputKey.Invalid);
```
we should maybe already  create an issue for this... this will definitely be problem in future... that is a direct reference which could be very problematic... I don't see a problem for the pr now, but we should tackle this soon we can maybe utilize the input system on our own

hunharibo replied:
1.6.1 has a whole new system for custom keybinds. We should wait with this for that.

Z3rca replied:
yeah right. maybe 1.6.3 or .1.6.4. will be finnally our jump ^^'

Badulundabad replied:
You're right, will be waiting for 1.6.1

Z3rca commented on CSharpSourceCode/Abilities/Crosshairs/AbilityCrosshair.cs:
```diff
@@ -0,0 +1,180 @@
+    /// <summary>
+    /// Base class for all crosshairs
+    /// </summary>
+    public abstract class AbilityCrosshair : IDisposable
```
I believe it might be a better option to call it a  ProjectedAbilityCrosshair, to avoid confusions

Z3rca commented on CSharpSourceCode/Abilities/Crosshairs/TargetedAOECrosshair.cs:
```diff
@@ -0,0 +1,83 @@
+            UpdateFrame();
+            if (CollidedAgents != null)
+            {
+                previousAgents = (Agent[])CollidedAgents.Clone();
```
what exactly does here clone do?

Badulundabad replied:
i need to get copy of CollidedAgents array, not reference. Maybe it's excess, idk

Z3rca commented on CSharpSourceCode/Abilities/Scripts/AbilityScript.cs:
```diff
@@ -30,7 +31,9 @@ internal void SetAgent(Agent agent)
         protected virtual bool ShouldMove()
         {
-            return _ability.Template.AbilityEffectType != AbilityEffectType.TargetedStaticAOE && _ability.Template.AbilityEffectType != AbilityEffectType.CenteredStaticAOE;
+            return _ability.Template.AbilityEffectType != AbilityEffectType.TargetedStaticAOE &&
```
not for now, but potentially we should work with flags... it reads really bad

Badulundabad replied:
yeah, i agree

Z3rca reviewed (commented):
I added some commits. in principle a good thing, I would get rid of some magic numbers, and I would like to know why you did some of the changes I marked

Z3rca reviewed (approved):
for speeding up the process, I approve, but please act, or note the comments, and we talk about such things later :)

Badulundabad committed:
Made a lot of changes. Read description.

- Removed AOECrosshairs abstract class;
- Adjusted CenteredAOECrosshair;
- Fixed game crashing where player is trying to walk around the city/castle/village;
- AbilityTargetType of HealingAOE was changed to Allies;
- Made some small changes.

Badulundabad committed:
Made some small changes
